internal/config/migrate.go

```diff
@@ -0,0 +1,56 @@
+package config
+
+import (
+	"fmt"
+
+	ghConfig "github.com/cli/go-gh/v2/pkg/config"
+)
+
+//go:generate moq -rm -out migration_mock.go . Migration
+
+// Migration is the interace that config migrations must implement.
+//
+// Migrations will receive a copy of the config, and should modify that copy
+// as necessary. After migration has completed, the modified config contents
+// will be used.
+//
+// The calling code is expected to verify that the current version of the config
+// matches the PreVersion of the migration before calling Do, and will set the
+// config version to the PostVersion after the migration has completed successfully.
+type Migration interface {
+	// PreVersion is the required config version for this to be applied
+	PreVersion() string
+	// PostVersion is the config version that must be applied after migration
+	PostVersion() string
+	// Do is expected to apply any necessary changes to the config in place
+	Do(*ghConfig.Config) error
+}
+
+func Migrate(c *ghConfig.Config, m Migration) error {
```

Comment:
How would you feel about making this a method?

```go
func (c *Config) Migrate(m Migration) error
```

To me it feels like since we've moved it up to `cli` it naturally belongs on the `cli Config` even though it is a bit leaky about the internal structure that seems unavoidable with the interface that actually gets exposed right now.

The other thing I like about having the `cli Config` as a method receiver is that I _think_ in the future there might well be value in making the `Keyring` an injectable dependency and then it's a first class object for migrations rather than a kind of hidden dependency.

Reply:
Yeah I am on board with this. I thought about making this change already but was worried about exposing the internal structure.

It would be nice to have `keyring` as an injectable dependency, something we can work towards in the future.